pkg/internal/apis/acme/validation/challenge.go

```diff
@@ -23,21 +23,23 @@ import (
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/util/validation/field"
 
+	"github.com/jetstack/cert-manager/pkg/internal/api/validation"
 	cmacme "github.com/jetstack/cert-manager/pkg/internal/apis/acme"
 )
 
-func ValidateChallengeUpdate(_ *admissionv1.AdmissionRequest, oldObj, newObj runtime.Object) field.ErrorList {
+func ValidateChallengeUpdate(_ *admissionv1.AdmissionRequest, oldObj, newObj runtime.Object) (field.ErrorList, validation.WarningList) {
 	old, ok := oldObj.(*cmacme.Challenge)
 	new := newObj.(*cmacme.Challenge)
 	// if oldObj is not set, the Update operation is always valid.
 	if !ok || old == nil {
-		return nil
+		return nil, nil
 	}
 
+	var warnings validation.WarningList
```

Comment:
> Would it be better to not have warnings defined here, and instead just return el, nil at the end so we have one less line of code to reason about?

Agreed, I've changed that.

> Similarly in tests for these functions, we could simplify
In the tests, I would rather explicitly check for it (i.e to catch a case where an expected warning is not returned sometime in future). Happy to be convinced otherwise.

> There could be a case for maybe having ValidateFunc and ValidateUpdateFunc return, say, a ValidationResponse struct. I think that could make a lot of the testing logic cleaner

I don't feel strongly either way. I guess if it was expected that at some point in the future there will be more values to be returned by those functions than a struct would definitely make sense. At the moment it almost feels like having these two values directly is simpler. Happy to put them in a struct or not.

Reply:
> I guess if it was expected that at some point in the future there will be more values to be returned by those functions than a struct would definitely make sense

That's a really good point, yeah. I think for now it's totally fine to have the two return values, as we've got it now; the difference between that and a struct is so minimal :+1: